### ISSUE TITLE
Mode 2 - Rootster 604 Unloader and AI 

### ISSUE THREAD

**Issue by GeorgeEarslight** [Courseplay/courseplay#5789]
Hello guys, hope you're doing great.

I've run into an issue where the Mode 2 driver is stopping right before the Rootster 604 trailer beet harvester.
From the log, it seems the AI doesn't allow it to drive further. I've tried various offsets without any success.
The map is Sandy Bay, I've included channel 3,4,12 full log, log snippet, screenshots, and savegame.
The savegame is stripped down from my main (heavily modded) game.

[savegame3.zip](https://github.com/Courseplay/courseplay/files/4983065/savegame3.zip)
[log_full.txt](https://github.com/Courseplay/courseplay/files/4983059/log_full.txt)
[log_snippet.txt](https://github.com/Courseplay/courseplay/files/4983060/log_snippet.txt)
![1](https://user-images.githubusercontent.com/30686125/88564975-9c78d080-d03c-11ea-9a66-2f932144693d.png)
![2](https://user-images.githubusercontent.com/30686125/88564983-a0a4ee00-d03c-11ea-884c-3c42e49b7377.png)
![3](https://user-images.githubusercontent.com/30686125/88564994-a4d10b80-d03c-11ea-9ea6-a43505eb8b7f.png)

**Comment by Tensuko** [comment]
I wonder a bit about the points, that it gets into the state waiting_for_combine_to_call.
It looks like AI colli stopped it.
In your logfile it looks all smoth to me.
Last entry:
2020-07-27 18:55 :57 [dbg4 lp60101] 5613 (#2):  -> 7722 (#1): combine empty and moving forward
2020-07-27 18:55 :57 [dbg4 lp60101] 5613 (#2):  -> N/A: setNewOnFieldState: WAITING_FOR_COMBINE_TO_CALL

**Comment by GeorgeEarslight** [comment]
Hey, @Tensuko thanks for looking into it.

`[dbg4 lp60100] 5613 (#2): Speed = 24.0, gx=-15.6 gz=19.2 l=24.8 ax=-5.3 az=6.5 allowed=false fwd=true`
`[dbg3 lp60100] 5613 (#2): has 0 colliding object(s)`

Didn't make sense to me that allowed=false but has 0 colliding objects, that's why I've mentioned the AI initially.
Do you want me to open another channel or change any options?
I might try the debugger and see if I can get something there.

**Comment by pvaiko** [comment]
@GeorgeEarslight glad to see you again.

No, its what it says: "combine empty and moving forward", it thinks the combine is empty and lets it go. Can you check one more thing, does it work if you switch the mode for the harvester to Mode 5 and then back?

**Comment by GeorgeEarslight** [comment]
Hey @pvaiko same here, glad you're keeping up with the load.

I'll check when I get back home and report. 
I'll also try to check more logs on the "combine" tractor.

**Comment by GeorgeEarslight** [comment]
@pvaiko I've run on debug, you don't have a `getCurrentDischargeNode` on the combine class, you're looking at the tractor instead of the trailer. (mode 5 didn't work).

```lua
function CombineUnloadManager:getCombinesFillLevelPercent(combine)
	if not combine.getCurrentDischargeNode then
		-- TODO: cotton harvesters for example don't have one...
		return 0
	end
	local dischargeNode = combine:getCurrentDischargeNode()
	return combine:getFillUnitFillLevelPercentage(dischargeNode.fillUnitIndex)*100
end
```
![image](https://user-images.githubusercontent.com/30686125/88694258-f646cc80-d108-11ea-894e-1a7fe2ff8164.png)

**Comment by pvaiko** [comment]
Thanks for the troubleshooting, that CombineUnloadManager is still a mess, I'm not done with cleaning it up, don't even know why don't we go directly to the combine to ask for the fill level.

**Comment by GeorgeEarslight** [comment]
Well somewhere there you've got knowledge of the trailed harvester otherwise, you wouldn't add it on the combine list. 
I'm gonna dig a bit more and report or open a pull.

**Comment by pvaiko** [comment]
Here:
https://github.com/Courseplay/courseplay/blob/00f6ec82d045c0b6a29f12a3b181be68be8873b8/CombineUnloadManager.lua#L74

**Comment by GeorgeEarslight** [comment]
The problem is here, combine in pair doesn't carry `getCurrentDischargeNode` when non-self-propelled, no idea why. It does on normal harvesters.
The encouraging thing is that `attributes.combineObject` does carry it, so it might be able to be replaced at least for the offending attribute updates. 
https://github.com/Courseplay/courseplay/blob/00f6ec82d045c0b6a29f12a3b181be68be8873b8/CombineUnloadManager.lua#L287-L306

**Comment by GeorgeEarslight** [comment]
@pvaiko 
I've managed to make some minor changes and point to the actual combine (trailed or otherwise) even if calling the vehicle, but unfortunately there are too many calls to the combine (mostly from the unloader) while assuming it is also a vehicle.
I think this would require to refactor everything combine related to deal with them on a more abstract level without assuming they're vehicles, unless you could think of something else.

**Comment by pvaiko** [comment]
I think all we need is to implement this in a proper OO way and just add a function to the CombineAIDriver to get the fill level and use that everywhere consistently.

**Comment by GeorgeEarslight** [comment]
I've already done but had additional issues because of the vehicle and combine crossover. I'll update later today, perhaps it's my own misunderstanding of how the driver works.

**Comment by pvaiko** [comment]
CombineAIDriver.combine always points to the Combine specialization, even if it is a towed or mounted implement.

**Comment by GeorgeEarslight** [comment]
There are cases where the vehicle is added in lists
https://github.com/Courseplay/courseplay/blob/00f6ec82d045c0b6a29f12a3b181be68be8873b8/CombineAIDriver.lua#L147
https://github.com/Courseplay/courseplay/blob/00f6ec82d045c0b6a29f12a3b181be68be8873b8/CombineUnloadManager.lua#L68-L78
and then used for some reason as a combine spec
https://github.com/Courseplay/courseplay/blob/00f6ec82d045c0b6a29f12a3b181be68be8873b8/CombineUnloadManager.lua#L290-L295

Even here combine points to the vehicle.
https://github.com/Courseplay/courseplay/blob/8531d9d7a8e4001cfd84e7518cc939d811c1e515/Events/UnloaderEvents.lua#L19-L24
from here
https://github.com/Courseplay/courseplay/blob/8531d9d7a8e4001cfd84e7518cc939d811c1e515/Events/UnloaderEvents.lua#L78-L83
from here
https://github.com/Courseplay/courseplay/blob/00f6ec82d045c0b6a29f12a3b181be68be8873b8/CombineUnloadManager.lua#L131-L140
finally because here the list I've mentioned earlier contains the vehicle and not the combine.
https://github.com/Courseplay/courseplay/blob/00f6ec82d045c0b6a29f12a3b181be68be8873b8/CombineUnloadManager.lua#208-L226

I've already tried to add the combine instead of the vehicle here
```lua
g_combineUnloadManager:addCombineToList(self.vehicle, self) 
```

But there are other places in the code where the actual vehicle is required as it's being stopped, started etc.
Hence the helper in my push is just a copy of the code you were using to get the combine spec from the vehicle when initializing the combine driver.

**Comment by GeorgeEarslight** [comment]
Solved by  62dee33.
